FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+- there should be the possibility to provide additional options to the indexes that support specific selection features (e.g., Scikit-learn's `BallTree`'s `dualtree` query option to boost performance).
+  - the best API is not trivial here, since `.sel()` may accept indexers passed to several indexes (which should still be supported for convenience and compatibility), and indexes may have similar options with different semantics
+  - we could introduce a new parameter like `index_options: Dict[XarrayIndex, Dict[str, Any]]` to pass options grouped by index
+- the `method` and `tolerance` parameters are specific to `pandas.Index` and would not be supported by all indexes: probably best is to eventually pass those arguments as `index_options`
+- the list valid indexer types might be extended in order to support new ways of indexing data, e.g., unordered selection of all points within a given range
+  - alternatively, we could reuse existing indexer types with different semantics depending on the index, e.g., using `slice(min, max, None)` for unordered range selection
+
+With the new data model proposed here, once ambiguous situation may occur when indexers are given for several coordinates that share the same dimension but not the same index, e.g., from the example in [Section 1](#1-data-model):
+
+```python
+da.sel(x=..., y=..., lat=..., lon=...)
+```
+
+The easiest solution for this situation would be to raise an error. Alternatively, we could introduce a new parameter to specify how to combine the resulting integer indexers (i.e., union vs intersection), although this could already be achieved by chaining `.sel()` calls or combining `.sel()` with `.merge()` (it may or may not be straightforward).
+
+### 2.6 Using indexes for data alignment
+
+Another main use if indexes is data alignment in various operations. Some considerations regarding alignment and flexible indexes:
+
+- support for alignment should probably be optional for an `XarrayIndex` subclass.
+  - like `pandas.Index`, the index wrapper classes that support it should implement `.equals()`, `.union()` and/or `.intersection()`
+  - support might be partial if that makes sense (outer, inner, left, right, exact...).
+  - index equality might involve more than just the labels: for example a spatial index might be used to check if the coordinate system (CRS) is identical for two sets of coordinates
+  - some indexes might implement inexact alignment, like in [#4489](https://github.com/pydata/xarray/pull/4489) or a `KDTree` index that selects nearest-neighbors within a given tolerance
+  - alignment may be "multi-dimensional", i.e., the `KDTree` example above vs. dimensions aligned independently of each other
+- we need to decide what to do when one dimension has more than one index that supports alignment
```

Comment:
Here and elsewhere it sounds like it could be useful to define an index precedence order.

Reply:
Yes this could be a solution to all situations where the index to pick is ambiguous (although it is not yet clear to me how often such situation could happen). This would introduce more complexity, though:

- we need to extend the API so that users can define this order
- we need to ensure strict precedence order
- does it make sense to use the same precedence order for all operations? (e.g., alignment vs. selection)